File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.go
type TrieNode struct {
    children map[rune]*TrieNode
    isEnd    bool
}

type Trie struct {
    root *TrieNode
}

func Constructor() Trie {
    return Trie{
        root: &TrieNode{
            children: make(map[rune]*TrieNode),
        },
    }
}

func (t *Trie) Insert(word string) {
    node := t.root
    
    for _, char := range word {
        if _, exists := node.children[char]; !exists {
            node.children[char] = &TrieNode{
                children: make(map[rune]*TrieNode),
            }
        }
        node = node.children[char]
    }
    
    node.isEnd = true
}

func (t *Trie) Search(word string) bool {
    node := t.traverse(word)
    return node != nil && node.isEnd
}

func (t *Trie) StartsWith(prefix string) bool {
    return t.traverse(prefix) != nil
}

func (t *Trie) traverse(str string) *TrieNode {
    node := t.root
    
    for _, char := range str {
        if _, exists := node.children[char]; !exists {
            return nil
        }
        node = node.children[char]
    }
    
    return node
}

File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.js
var Trie = function() {
    this.root = {};
};

Trie.prototype.insert = function(word) {
    let node = this.root;
    
    for (const char of word) {
        if (!node[char]) {
            node[char] = {};
        }
        node = node[char];
    }
    
    node.isEnd = true;
};

Trie.prototype.search = function(word) {
    const node = this._traverse(word);
    return node !== null && node.isEnd === true;
};

Trie.prototype.startsWith = function(prefix) {
    return this._traverse(prefix) !== null;
};

// 헬퍼: prefix까지 탐색
Trie.prototype._traverse = function(str) {
    let node = this.root;
    
    for (const char of str) {
        if (!node[char]) {
            return null;
        }
        node = node[char];
    }
    
    return node;
};